src/link/script.y

```diff
@@ -0,0 +1,479 @@
+}
+
+static bool isWhiteSpace(LexerStackEntry::int_type c) {
+	return c == ' ' || c == '\t';
+}
+
+static bool isNewline(LexerStackEntry::int_type c) {
+	return c == '\r' || c == '\n';
+}
+
+static bool isIdentChar(LexerStackEntry::int_type c) {
+	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
+}
+
+static bool isHexDigit(LexerStackEntry::int_type c) {
+	return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
+}
+
+static uint8_t parseHexDigit(LexerStackEntry::int_type c) {
+	if (c >= '0' && c <= '9') {
+		return c - '0';
+	} else if (c >= 'A' && c <= 'F') {
+		return c - 'A' + 10;
+	} else if (c >= 'a' && c <= 'f') {
+		return c - 'a' + 10;
+	} else {
+		unreachable_();
+	}
+}
+
+yy::parser::symbol_type yylex(void) {
+try_again: // Can't use a `do {} while(0)` loop, otherwise compilers (wrongly) think it can end.
+	auto &context = lexerStack.back();
+	auto c = context.file.sbumpc();
+
+	// First, skip leading whitespace.
+	while (isWhiteSpace(c)) {
+		c = context.file.sbumpc();
+	}
+	// Then, skip a comment if applicable.
+	if (c == ';') {
+		while (!isNewline(c)) {
+			c = context.file.sbumpc();
+		}
+	}
+
+	// Alright, what token should we return?
+	if (c == LexerStackEntry::eof) {
+		// Basically yywrap().
+		lexerStack.pop_back();
+		if (!lexerStack.empty()) {
+			goto try_again;
+		}
+		return yy::parser::make_YYEOF();
+	} else if (isNewline(c)) {
+		// Handle CRLF.
+		if (c == '\r' && context.file.sgetc() == '\n') {
+			context.file.sbumpc();
+		}
+		return yy::parser::make_newline();
+	} else if (c == '"') {
+		std::string str;
+
+		for (c = context.file.sgetc(); c != '"'; c = context.file.sgetc()) {
+			if (c == LexerStackEntry::eof || isNewline(c)) {
+				scriptErrorSimple(context, "Unterminated string");
+				break;
+			}
+			context.file.sbumpc();
+			if (c == '\\') {
+				c = context.file.sgetc();
+				if (c == LexerStackEntry::eof || isNewline(c)) {
+					scriptErrorSimple(context, "Unterminated string");
+					break;
+				} else if (c == 'n') {
+					c = '\n';
+				} else if (c == 'r') {
+					c = '\r';
+				} else if (c == 't') {
+					c = '\t';
+				} else if (c != '\\' && c != '"') {
+					scriptError(context, "Cannot escape character %s", printChar(c));
+				}
+				context.file.sbumpc();
+			}
+			str.push_back(c);
+		}
+		context.file.sbumpc(); // Consume the closing quote.
+
+		return yy::parser::make_string(std::move(str));
+	} else if (c == '$') {
+		c = context.file.sgetc();
+		if (!isHexDigit(c)) {
+			scriptErrorSimple(context, "No hexadecimal digits found after '$'");
+			return yy::parser::make_number(0);
+		}
+
+		uint32_t number = parseHexDigit(c);
+		context.file.sbumpc();
+		for (c = context.file.sgetc(); isHexDigit(c); c = context.file.sgetc()) {
+			number = number * 16 + parseHexDigit(c);
+			context.file.sbumpc();
+		}
+		return yy::parser::make_number(number);
+	} else if (c == '%') {
+		c = context.file.sgetc();
+		if (!(c >= '0' && c <= '1')) {
```

Comment:
```suggestion
		if (c != '0' && c != '1') {
```

Reply:
I preferred that for symmetry with the other branches. I'm not particularly attached to it, though.